npm/vite-dev-server/src/plugins/cypress.ts

```diff
@@ -133,13 +133,15 @@ export const Cypress = (
         // as soon as we find one of the specs, we trigger the re-run of tests
         for (const mod of moduleImporters.values()) {
           debug('handleHotUpdate - mod.file', mod.file)
+          // always re-run on supportFile changed
           if (mod.file === supportFilePath) {
-            debug('handleHotUpdate - support compile success')
-            devServerEvents.emit('dev-server:compile:success')
```

Comment:
I was under the impression that `devServerEvents.emit('dev-server:compile:success')` is similar to `server.ws.send( type: 'full-reload' })` as it destroys the AUT and creates a new one.

Seems like we have two mechanisms here for causing a reload, the `full-reload` and the `dev-server:compile:success` event. Do we know why the `dev-server:compile:success` wasn't working in this case?

Reply:
I _think_ `dev-server:compile:success` just causes the spec to re-run (using any updated code via HMR). `full-reload` is a Vite client API which tells Vite to actually go and get the latest resources in their entirety (since HMR isn't working for the stylesheets).

So this change basically means we aren't using HMR for Vite anymore, but doing a full reload. This seems more correct imo - fresh slate every test WDYT?

HMR will be useful if we have a "workbench" mode one day like Storybook... but not so much for tests.

Reply:
Hmmm I think the problem might be Vite caching, then? :thinking: 

`dev-server:compile:success` triggers a *re-run* but has nothing to do with fetching the latest data.

https://github.com/cypress-io/cypress/blob/develop/packages/app/src/runner/event-manager.ts#L153-L156

Right now it's 

- HMR -> get latest data (spec, component) -> `dev-server:compile:success` -> `rerun()`.

So the problem is Vite - I think it's cache is not getting updated, so it uses the stale one. I think for reliability we should probably just be using `full-reload` for Vite exclusively? It's more in line with the "fresh slate" and "test isolation" philosophy, for sure.

What do you think?

Reply:
I think you are right @ZachJW34, I will double down and find out why the HMR trigger doesn't work.

Reply:
I investigated. I think we are NOT using HMR what-so-ever right now. According to the [Vite HMR docs](https://vitejs.dev/guide/api-plugin.html#handlehotupdate).

> Return an empty array and perform complete custom HMR handling by sending custom events to the client

We currently return an empty array, but we don't do any custom HMR (like `import.meta.hot...`). The assets *are* rebuilt (the JS modules in memory) but we don't do any HMR updates. We just emit the `dev-server:compile:success`, which casuse Cypress to re-run the spec, which re-requests the assets.

We have a few options to fix this. 

- Just be to change `return []` to `return`. This seems to work just fine.
- use the new `reloadModule` API.

```
 /**
   * Triggers HMR for a module in the module graph. You can use the `server.moduleGraph`
   * API to retrieve the module to be reloaded. If `hmr` is false, this is a no-op.
   */
  reloadModule(module: ModuleNode): Promise<void>
```

I tried this, I got it working with `if (mod.file.endsWith('css')) { reloadModule(mod) }`. Although this probably isn't what we want - it's more for forcing a virtual module to reload. It's also not available until Vite 3.2.0 - which seems to NOT be a straight forward upgrade :(

Reply:
I did some poking around of [Vite's HMR](https://github.com/vitejs/vite/blob/main/packages/vite/src/node/server/hmr.ts#L100) and I can definitely see how `return []` could cause some cache issues. Looks like returning an empty array essentially tells Vite that there is no new module and the previous module isn't invalidated. Maybe this is only a problem with virtual modules which is why we are seeing this with Tailwind.

I'm cool with the fix you have here! Not sure why the vite-dev-server tests crapped out but I'll rerun them.

Reply:
Maybe we move to `reloadModule` once we update to Vite 3.2. That's supposed to be for virtual modules, like Tailwind (I assume this is how it's implemented, anyway).